components/raftstore/src/store/worker/read.rs

```diff
@@ -1546,135 +1627,274 @@ mod tests {
+            1,
+            term6,
+            pr_ids1,
+            epoch13.clone(),
+            store_meta.clone(),
+        );
+        let leader1 = prs1[0].clone();
+
+        // Register region2
+        let pr_ids2 = vec![22, 33, 44];
+        let prs2 = new_peers(store_id, pr_ids2.clone());
+        prepare_read_delegate(store_id, 2, term6, pr_ids2, epoch13.clone(), store_meta);
+        let leader2 = prs2[0].clone();
+
+        reader.kv_engine.put(b"a1", b"val1").unwrap();
+
+        let mut cmd = RaftCmdRequest::default();
+        let mut header = RaftRequestHeader::default();
+        header.set_region_id(1);
+        header.set_peer(leader1);
+        header.set_region_epoch(epoch13.clone());
+        header.set_term(term6);
+        cmd.set_header(header);
+        let mut req = Request::default();
+        req.set_cmd_type(CmdType::Snap);
+        cmd.set_requests(vec![req].into());
+
+        let task = RaftCommand::<KvTestSnapshot>::new(
+            cmd.clone(),
+            Callback::read(Box::new(move |resp: ReadResponse<KvTestSnapshot>| {
+                let snap = resp.snapshot.unwrap();
+                assert_eq!(snap.get_region().id, 1);
+                assert_eq!(
+                    snap.get_snapshot().get_value(b"a1").unwrap().unwrap(),
+                    b"val1"
+                );
+            })),
+        );
+
+        // First request will not hit cache
+        let read_id = Some(ThreadReadId::new());
+        must_not_redirect_with_read_id(&mut reader, &rx, task, read_id.clone());
         assert_eq!(
-            b"val1".to_vec(),
-            *snapshot.get_value(b"a1").unwrap().unwrap()
+            TLS_LOCAL_READ_METRICS.with(|m| m.borrow_mut().local_executed_snapshot_cache_hit.get()),
+            0
+        );
+        assert_eq!(
+            TLS_LOCAL_READ_METRICS.with(|m| m.borrow_mut().local_executed_requests.get()),
+            1
         );
 
-        assert!(snap_cache.as_ref().is_some());
+        // Change the value of a1, but if cache hit, it still reader "val1"
+        reader.kv_engine.put(b"a1", b"val2").unwrap();
+
+        let mut header = RaftRequestHeader::default();
+        header.set_region_id(2);
+        header.set_peer(leader2);
+        header.set_region_epoch(epoch13);
+        header.set_term(term6);
+        cmd.set_header(header);
+        let task = RaftCommand::<KvTestSnapshot>::new(
+            cmd.clone(),
+            Callback::read(Box::new(move |resp: ReadResponse<KvTestSnapshot>| {
+                let snap = resp.snapshot.unwrap();
+                assert_eq!(snap.get_region().id, 2);
+                assert_eq!(
+                    snap.get_snapshot().get_value(b"a1").unwrap().unwrap(),
+                    b"val1"
+                );
+            })),
+        );
+        must_not_redirect_with_read_id(&mut reader, &rx, task, read_id);
+        assert_eq!(
+            TLS_LOCAL_READ_METRICS.with(|m| m.borrow_mut().local_executed_snapshot_cache_hit.get()),
```

Comment:
Or you can just check if `Arc` is the same by `Arc::ptr_eq`.

Reply:
Done